File: 532.K-diffPairsInAnArray/diffPair.cc
class Solution {
public:
    int findPairs(vector<int>& nums, int k) {
        if (k < 0) {
            return 0;
        }

        map<int,int> m;
        for (int i = 0; i < int(nums.size()); i++) {
            map<int,int>::iterator it = m.find(nums[i]);
            if (it == m.end()) {
                m[nums[i]] = 0;
            }
            m[nums[i]]++;
        }

        int count = 0;

        if (k > 0) {
            for (map<int,int>::iterator it = m.begin(); it != m.end(); ++it) {
                map<int,int>::iterator hit = m.find(it->first + k);
                if (hit != m.end()) {
                    count++;
                }
            }
        }
        else {
            for (map<int,int>::iterator it = m.begin(); it != m.end(); ++it) {
                if (it->second > 1) {
                    count++;
                }
            }
        }

        return count;
    }
};

File: 532.K-diffPairsInAnArray/DiffPair.java
public class DiffPair {
    public int findPairs(int[] nums, int k) {
        if (k < 0) {
            return 0;
        }

        Map<Integer, Integer> map = new HashMap<>();

        for (int n : nums) {
            if (!map.containsKey(n)) {
                map.put(n, 0);
            }
            map.put(n, map.get(n)+1);
        }

        int count = 0;

        if (k > 0) {
            for (int n : map.keySet()) {
                if (map.containsKey(n + k)) {
                    count++;
                }
            }
        }
        else {
            for (int v : map.values()) {
                if (v > 1) {
                    count++;
                }
            }
        }

        return count;
    }
}
